File: cpp_app/diagnosis.cpp
// Imperative disease prediction rules
#include "diagnosis.h"
#include <algorithm>
#include <set>

using namespace std;

// check if a symptom exists in the list
bool hasSymptom(const vector<string>& symptoms, const string& symptom) {
    return find(symptoms.begin(), symptoms.end(), symptom) != symptoms.end();
}

// check if multiple symptoms exist
bool hasSymptoms(const vector<string>& symptoms, const vector<string>& requiredSymptoms) {
    for (const auto& symptom : requiredSymptoms) {
        if (!hasSymptom(symptoms, symptom)) {
            return false;
        }
    }
    return true;
}

vector<string> predictDiseases(const vector<string>& symptoms) {
    set<string> diseases; // Using set to avoid duplicates

    // Rule 1: Common Cold
    if (hasSymptoms(symptoms, {"runny nose", "sore throat"}) ||
        hasSymptoms(symptoms, {"runny nose", "cough"}) ||
        hasSymptoms(symptoms, {"sore throat", "cough"}) ||
        (hasSymptom(symptoms, "runny nose") && hasSymptom(symptoms, "headache"))) {
        diseases.insert("Common Cold");
    }

    // Rule 2: Flu (Influenza)
    if (hasSymptoms(symptoms, {"fever", "cough"}) ||
        hasSymptoms(symptoms, {"fever", "muscle aches"}) ||
        hasSymptoms(symptoms, {"fever", "fatigue", "headache"}) ||
        hasSymptoms(symptoms, {"fever", "chills"}) ||
        hasSymptoms(symptoms, {"muscle aches", "fatigue", "headache"})) {
        diseases.insert("Flu");
    }

    // Rule 3: COVID-19
    if (hasSymptoms(symptoms, {"fever", "cough", "loss of taste"}) ||
        hasSymptoms(symptoms, {"fever", "shortness of breath"}) ||
        hasSymptoms(symptoms, {"loss of taste", "loss of smell"}) ||
        hasSymptoms(symptoms, {"fever", "fatigue", "muscle aches"}) ||
        hasSymptoms(symptoms, {"cough", "loss of taste"}) ||
        hasSymptoms(symptoms, {"cough", "loss of smell"}) ||
        hasSymptoms(symptoms, {"fever", "headache", "sore throat"})) {
        diseases.insert("COVID-19");
    }

    // Rule 4: Pneumonia
    if (hasSymptoms(symptoms, {"fever", "cough", "shortness of breath"}) ||
        hasSymptoms(symptoms, {"chest pain", "cough", "fever"}) ||
        hasSymptoms(symptoms, {"shortness of breath", "chest pain"}) ||
        hasSymptoms(symptoms, {"fever", "chills", "shortness of breath"})) {
        diseases.insert("Pneumonia");
    }

    // Rule 5: Gastroenteritis (Stomach Flu)
    if (hasSymptoms(symptoms, {"nausea", "vomiting", "diarrhea"}) ||
        hasSymptoms(symptoms, {"nausea", "diarrhea"}) ||
        hasSymptoms(symptoms, {"vomiting", "diarrhea"}) ||
        hasSymptoms(symptoms, {"nausea", "vomiting"}) ||
        (hasSymptom(symptoms, "diarrhea") && hasSymptom(symptoms, "fever"))) {
        diseases.insert("Gastroenteritis");
    }

    // Rule 6: Migraine
    if (hasSymptoms(symptoms, {"headache", "nausea"}) ||
        hasSymptoms(symptoms, {"headache", "dizziness"}) ||
        (hasSymptom(symptoms, "headache") && hasSymptom(symptoms, "vomiting"))) {
        diseases.insert("Migraine");
    }

    // Rule 7: Allergic Reaction
    if (hasSymptoms(symptoms, {"rash", "runny nose"}) ||
        hasSymptoms(symptoms, {"rash", "sore throat"}) ||
        (hasSymptom(symptoms, "rash") && hasSymptom(symptoms, "shortness of breath"))) {
        diseases.insert("Allergic Reaction");
    }

    // Rule 8: Strep Throat
    if (hasSymptoms(symptoms, {"sore throat", "fever"}) &&
        !hasSymptom(symptoms, "runny nose") &&
        !hasSymptom(symptoms, "cough")) {
        diseases.insert("Strep Throat");
    }

    // Rule 9: Bronchitis
    if (hasSymptoms(symptoms, {"cough", "chest pain"}) ||
        hasSymptoms(symptoms, {"cough", "fatigue"}) ||
        (hasSymptom(symptoms, "cough") && hasSymptom(symptoms, "shortness of breath"))) {
        diseases.insert("Bronchitis");
    }

    // Rule 10: Food Poisoning
    if (hasSymptoms(symptoms, {"nausea", "vomiting"}) ||
        hasSymptoms(symptoms, {"diarrhea", "nausea"}) ||
        hasSymptoms(symptoms, {"vomiting", "diarrhea", "fever"})) {
        diseases.insert("Food Poisoning");
    }

    // Rule 11: Sinusitis
    if (hasSymptoms(symptoms, {"headache", "runny nose"}) ||
        hasSymptoms(symptoms, {"headache", "sore throat", "runny nose"}) ||
        (hasSymptom(symptoms, "headache") && hasSymptom(symptoms, "fever") && hasSymptom(symptoms, "runny nose"))) {
        diseases.insert("Sinusitis");
    }

    // Rule 12: Asthma Attack
    if (hasSymptoms(symptoms, {"shortness of breath", "cough"}) ||
        (hasSymptom(symptoms, "shortness of breath") && hasSymptom(symptoms, "chest pain"))) {
        diseases.insert("Asthma");
    }

    // Rule 13: Anxiety/Panic Attack
    if (hasSymptoms(symptoms, {"shortness of breath", "dizziness"}) ||
        hasSymptoms(symptoms, {"chest pain", "dizziness"}) ||
        hasSymptoms(symptoms, {"nausea", "dizziness", "shortness of breath"})) {
        diseases.insert("Anxiety/Panic Attack");
    }

    // Rule 14: Dehydration
    if (hasSymptoms(symptoms, {"dizziness", "fatigue"}) ||
        hasSymptoms(symptoms, {"headache", "dizziness", "fatigue"})) {
        diseases.insert("Dehydration");
    }

    // Rule 15: Arthritis/Joint Issues
    if (hasSymptom(symptoms, "joint pain")) {
        if (hasSymptom(symptoms, "fever")) {
            diseases.insert("Inflammatory Arthritis");
        } else {
            diseases.insert("Arthritis");
        }
    }

    // Single symptom conditions
    if (hasSymptom(symptoms, "fever") && symptoms.size() == 1) {
        diseases.insert("Fever (Unknown Cause)");
    }

    if (hasSymptom(symptoms, "headache") && symptoms.size() == 1) {
        diseases.insert("Tension Headache");
    }

    if (hasSymptom(symptoms, "rash") && symptoms.size() == 1) {
        diseases.insert("Skin Condition");
    }

    // Convert set to vector
    return vector<string>(diseases.begin(), diseases.end());
}

File: prolog_version/diagnosis.pl
% Prolog disease prediction rules
% Main diagnosis system using declarative programming

% Include the knowledge base
:- consult(knowledge_base).

% Helper predicates

% Check if a symptom is in the patient's symptom list
has_patient_symptom(PatientSymptoms, Symptom) :-
    member(Symptom, PatientSymptoms).

% Check if patient has all symptoms in a list
has_all_symptoms(_, []).
has_all_symptoms(PatientSymptoms, [Symptom|Rest]) :-
    has_patient_symptom(PatientSymptoms, Symptom),
    has_all_symptoms(PatientSymptoms, Rest).

% Check if patient has any of the symptoms in a list
has_any_symptom(PatientSymptoms, [Symptom|_]) :-
    has_patient_symptom(PatientSymptoms, Symptom).
has_any_symptom(PatientSymptoms, [_|Rest]) :-
    has_any_symptom(PatientSymptoms, Rest).

% Check if patient does NOT have a symptom
not_has_symptom(PatientSymptoms, Symptom) :-
    \+ has_patient_symptom(PatientSymptoms, Symptom).

% Count symptoms for a disease that patient has
count_matching_symptoms(PatientSymptoms, Disease, Count) :-
    findall(Symptom, (has_symptom(Disease, Symptom), has_patient_symptom(PatientSymptoms, Symptom)), MatchingSymptoms),
    length(MatchingSymptoms, Count).

% Disease prediction rules using pattern matching and logical conditions

% Rule 1: Common Cold
possible_disease(PatientSymptoms, common_cold) :-
    (has_all_symptoms(PatientSymptoms, [runny_nose, sore_throat]);
     has_all_symptoms(PatientSymptoms, [runny_nose, cough]);
     has_all_symptoms(PatientSymptoms, [sore_throat, cough]);
     has_all_symptoms(PatientSymptoms, [runny_nose, headache])).

% Rule 2: Flu
possible_disease(PatientSymptoms, flu) :-
    (has_all_symptoms(PatientSymptoms, [fever, cough]);
     has_all_symptoms(PatientSymptoms, [fever, muscle_aches]);
     has_all_symptoms(PatientSymptoms, [fever, fatigue, headache]);
     has_all_symptoms(PatientSymptoms, [fever, chills]);
     has_all_symptoms(PatientSymptoms, [muscle_aches, fatigue, headache])).

% Rule 3: COVID-19
possible_disease(PatientSymptoms, covid19) :-
    (has_all_symptoms(PatientSymptoms, [fever, cough, loss_of_taste]);
     has_all_symptoms(PatientSymptoms, [fever, shortness_of_breath]);
     has_all_symptoms(PatientSymptoms, [loss_of_taste, loss_of_smell]);
     has_all_symptoms(PatientSymptoms, [fever, fatigue, muscle_aches]);
     has_all_symptoms(PatientSymptoms, [cough, loss_of_taste]);
     has_all_symptoms(PatientSymptoms, [cough, loss_of_smell]);
     has_all_symptoms(PatientSymptoms, [fever, headache, sore_throat])).

% Rule 4: Pneumonia
possible_disease(PatientSymptoms, pneumonia) :-
    (has_all_symptoms(PatientSymptoms, [fever, cough, shortness_of_breath]);
     has_all_symptoms(PatientSymptoms, [chest_pain, cough, fever]);
     has_all_symptoms(PatientSymptoms, [shortness_of_breath, chest_pain]);
     has_all_symptoms(PatientSymptoms, [fever, chills, shortness_of_breath])).

% Rule 5: Gastroenteritis
possible_disease(PatientSymptoms, gastroenteritis) :-
    (has_all_symptoms(PatientSymptoms, [nausea, vomiting, diarrhea]);
     has_all_symptoms(PatientSymptoms, [nausea, diarrhea]);
     has_all_symptoms(PatientSymptoms, [vomiting, diarrhea]);
     has_all_symptoms(PatientSymptoms, [nausea, vomiting]);
     (has_patient_symptom(PatientSymptoms, diarrhea), has_patient_symptom(PatientSymptoms, fever))).

% Rule 6: Migraine
possible_disease(PatientSymptoms, migraine) :-
    (has_all_symptoms(PatientSymptoms, [headache, nausea]);
     has_all_symptoms(PatientSymptoms, [headache, dizziness]);
     (has_patient_symptom(PatientSymptoms, headache), has_patient_symptom(PatientSymptoms, vomiting))).

% Rule 7: Allergic Reaction
possible_disease(PatientSymptoms, allergic_reaction) :-
    (has_all_symptoms(PatientSymptoms, [rash, runny_nose]);
     has_all_symptoms(PatientSymptoms, [rash, sore_throat]);
     (has_patient_symptom(PatientSymptoms, rash), has_patient_symptom(PatientSymptoms, shortness_of_breath))).

% Rule 8: Strep Throat (specific exclusion rule)
possible_disease(PatientSymptoms, strep_throat) :-
    has_all_symptoms(PatientSymptoms, [sore_throat, fever]),
    not_has_symptom(PatientSymptoms, runny_nose),
    not_has_symptom(PatientSymptoms, cough).

% Rule 9: Bronchitis
possible_disease(PatientSymptoms, bronchitis) :-
    (has_all_symptoms(PatientSymptoms, [cough, chest_pain]);
     has_all_symptoms(PatientSymptoms, [cough, fatigue]);
     (has_patient_symptom(PatientSymptoms, cough), has_patient_symptom(PatientSymptoms, shortness_of_breath))).

% Rule 10: Food Poisoning
possible_disease(PatientSymptoms, food_poisoning) :-
    (has_all_symptoms(PatientSymptoms, [nausea, vomiting]);
     has_all_symptoms(PatientSymptoms, [diarrhea, nausea]);
     has_all_symptoms(PatientSymptoms, [vomiting, diarrhea, fever])).

% Rule 11: Sinusitis
possible_disease(PatientSymptoms, sinusitis) :-
    (has_all_symptoms(PatientSymptoms, [headache, runny_nose]);
     has_all_symptoms(PatientSymptoms, [headache, sore_throat, runny_nose]);
     (has_patient_symptom(PatientSymptoms, headache), has_patient_symptom(PatientSymptoms, fever), has_patient_symptom(PatientSymptoms, runny_nose))).

% Rule 12: Asthma
possible_disease(PatientSymptoms, asthma) :-
    (has_all_symptoms(PatientSymptoms, [shortness_of_breath, cough]);
     (has_patient_symptom(PatientSymptoms, shortness_of_breath), has_patient_symptom(PatientSymptoms, chest_pain))).

% Rule 13: Anxiety/Panic Attack
possible_disease(PatientSymptoms, anxiety_panic_attack) :-
    (has_all_symptoms(PatientSymptoms, [shortness_of_breath, dizziness]);
     has_all_symptoms(PatientSymptoms, [chest_pain, dizziness]);
     has_all_symptoms(PatientSymptoms, [nausea, dizziness, shortness_of_breath])).

% Rule 14: Dehydration
possible_disease(PatientSymptoms, dehydration) :-
    (has_all_symptoms(PatientSymptoms, [dizziness, fatigue]);
     has_all_symptoms(PatientSymptoms, [headache, dizziness, fatigue])).

% Rule 15: Arthritis conditions
possible_disease(PatientSymptoms, inflammatory_arthritis) :-
    has_patient_symptom(PatientSymptoms, joint_pain),
    has_patient_symptom(PatientSymptoms, fever).

possible_disease(PatientSymptoms, arthritis) :-
    has_patient_symptom(PatientSymptoms, joint_pain),
    not_has_symptom(PatientSymptoms, fever).

% Single symptom conditions
possible_disease(PatientSymptoms, fever_unknown_cause) :-
    has_patient_symptom(PatientSymptoms, fever),
    length(PatientSymptoms, 1).

possible_disease(PatientSymptoms, tension_headache) :-
    has_patient_symptom(PatientSymptoms, headache),
    length(PatientSymptoms, 1).

possible_disease(PatientSymptoms, skin_condition) :-
    has_patient_symptom(PatientSymptoms, rash),
    length(PatientSymptoms, 1).

% Main prediction predicate
predict_diseases(PatientSymptoms, Diseases) :-
    findall(Disease, possible_disease(PatientSymptoms, Disease), DiseaseList),
    sort(DiseaseList, Diseases).

% Alternative: Get diseases with confidence scores
predict_diseases_with_confidence(PatientSymptoms, DiseasesWithScores) :-
    findall(Disease-Score, (
        possible_disease(PatientSymptoms, Disease),
        count_matching_symptoms(PatientSymptoms, Disease, Score)
    ), DiseaseScorePairs),
    sort(2, @>=, DiseaseScorePairs, DiseasesWithScores).

% Query helpers for interactive use

% Simple query: What diseases match these symptoms?
% Usage: ?- diagnose([fever, cough], Diseases).
diagnose(PatientSymptoms, Diseases) :-
    predict_diseases(PatientSymptoms, Diseases).

% Detailed query with confidence scores
% Usage: ?- diagnose_detailed([fever, cough, headache], Results).
diagnose_detailed(PatientSymptoms, Results) :-
    predict_diseases_with_confidence(PatientSymptoms, Results).

% Interactive diagnosis - shows explanation
explain_diagnosis(PatientSymptoms) :-
    format('Patient symptoms: ~w~n', [PatientSymptoms]),
    predict_diseases(PatientSymptoms, Diseases),
    (   Diseases = []
    ->  format('No matching conditions found.~n')
    ;   format('Possible conditions:~n'),
        forall(member(Disease, Diseases), format('  - ~w~n', [Disease]))
    ).
